File: src/simulation/ai/llm/HTTPAgentClient.js
// HTTPAgentClient — pluggable AgentAdapter implementation that talks to the
// agent-bridge HTTP server (`server/agent-bridge.js`) over the
// `/api/agent/:agentId/*` route family.
//
// Unlike LLMClient (which is locked to the OpenAI-compatible /api/ai/* shape
// served by `server/ai-proxy.js`), HTTPAgentClient is channel-aware: each of
// the 4 channels (`environment-director` / `npc-policy` / `strategic-plan` /
// `colony-agent`) flows through the same `request(channel, payload, options)`
// entry point. The agent-bridge routes the request internally to whichever
// backend (LLMClient instance, FallbackAdapter, NoopAgentAdapter, vLLM, …)
// has been registered for that channel. This is the seam used by the E3
// cross-vendor matrix to assign different LLM vendors to different channels.
//
// Wire format (JSON over POST):
//   POST /api/agent/:agentId/decision
//     body  { channel, payload, schemaVersion, options }
//     reply { data, fallback?, usage?, latencyMs?, model?, error?, debug? }
//
// Failure policy mirrors AgentAdapter contract: never throw; on network /
// timeout / malformed response, return a DecisionResponse with
// `fallback: true` and a populated `error` so the caller can route to the
// deterministic policy fallback.

import { AgentAdapter, CHANNELS, SCHEMA_VERSION } from "./AgentAdapter.js";

const DEFAULT_BRIDGE_URL = "http://localhost:8788";
const DEFAULT_AGENT_ID = "default";
const DEFAULT_TIMEOUT_MS = 30_000;

function compactClientError(err) {
  const raw = String(err?.message ?? err ?? "unknown error")
    .replace(/\s+/g, " ")
    .trim();
  if (!raw) return "unknown error";
  const lower = raw.toLowerCase();
  if (lower.includes("timeout") || lower.includes("aborted")) return "request timeout";
  if (lower.includes("fetch failed") || lower.includes("econnref")) return "agent-bridge unreachable";
  return raw.slice(0, 200);
}

function nowMs() {
  return (typeof performance !== "undefined" && typeof performance.now === "function")
    ? performance.now()
    : Date.now();
}

function trimUrl(url) {
  return String(url ?? "").replace(/\/+$/, "");
}

function fallbackResponse(channel, error, latencyMs = 0, debug = null) {
  return {
    data: null,
    fallback: true,
    usage: { promptTokens: 0, completionTokens: 0, cachedTokens: 0 },
    latencyMs,
    model: "fallback",
    error: String(error ?? ""),
    debug: debug ?? { channel, source: "http-agent-client" },
  };
}

function isShapedDecisionResponse(obj) {
  // Accept anything that's an object with at least `data` (may be null) or
  // an explicit `fallback` flag. Reject arrays / primitives / null bodies.
  return Boolean(obj) && typeof obj === "object" && !Array.isArray(obj)
    && (Object.prototype.hasOwnProperty.call(obj, "data")
      || Object.prototype.hasOwnProperty.call(obj, "fallback"));
}

export class HTTPAgentClient extends AgentAdapter {
  /**
   * @param {object} opts
   * @param {string} [opts.agentBridgeUrl="http://localhost:8788"]
   * @param {string} [opts.agentId="default"]
   * @param {number} [opts.timeoutMs=30000]
   * @param {Record<string,string>} [opts.headers] — extra request headers
   * @param {typeof fetch} [opts.fetchImpl] — DI seam for tests
   */
  constructor(opts = {}) {
    super();
    this.agentBridgeUrl = trimUrl(opts.agentBridgeUrl ?? DEFAULT_BRIDGE_URL);
    this.agentId = String(opts.agentId ?? DEFAULT_AGENT_ID);
    this.timeoutMs = Math.max(500, Number(opts.timeoutMs ?? DEFAULT_TIMEOUT_MS) || DEFAULT_TIMEOUT_MS);
    this.headers = { ...(opts.headers ?? {}) };
    this.fetchImpl = opts.fetchImpl ?? globalThis.fetch;
    this.lastError = "";
    this.lastLatencyMs = 0;
    this.lastStatus = "unknown";
  }

  /**
   * @param {import("./AgentAdapter.js").Channel} channel
   * @param {object} payload
   * @param {object} [options] — { timeoutMs?, signal?, schemaVersion? }
   * @returns {Promise<import("./AgentAdapter.js").DecisionResponse>}
   */
  async request(channel, payload, options = {}) {
    if (!CHANNELS.includes(channel)) {
      return fallbackResponse(channel, `unknown channel: ${channel}`);
    }
    if (typeof this.fetchImpl !== "function") {
      return fallbackResponse(channel, "fetch unavailable in this runtime");
    }

    const url = `${this.agentBridgeUrl}/api/agent/${encodeURIComponent(this.agentId)}/decision`;
    const body = {
      channel,
      payload,
      schemaVersion: options.schemaVersion ?? SCHEMA_VERSION,
      options: {
        timeoutMs: options.timeoutMs ?? this.timeoutMs,
      },
    };

    const ctrl = new AbortController();
    const externalSignal = options.signal;
    if (externalSignal) {
      if (externalSignal.aborted) ctrl.abort(externalSignal.reason);
      else externalSignal.addEventListener("abort", () => ctrl.abort(externalSignal.reason), { once: true });
    }
    const timeoutMs = Math.max(100, Number(options.timeoutMs ?? this.timeoutMs) || this.timeoutMs);
    const timer = setTimeout(() => ctrl.abort("timeout"), timeoutMs);
    const started = nowMs();

    try {
      const resp = await this.fetchImpl(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          ...this.headers,
        },
        body: JSON.stringify(body),
        signal: ctrl.signal,
      });
      const latencyMs = nowMs() - started;
      this.lastLatencyMs = latencyMs;

      if (!resp.ok) {
        const text = await resp.text().catch(() => "");
        const short = text ? text.replace(/\s+/g, " ").trim().slice(0, 140) : "";
        this.lastStatus = "down";
        this.lastError = short ? `HTTP ${resp.status}: ${short}` : `HTTP ${resp.status}`;
        return fallbackResponse(channel, this.lastError, latencyMs);
      }

      let parsed;
      try {
        parsed = await resp.json();
      } catch (err) {
        this.lastStatus = "down";
        this.lastError = `malformed JSON: ${compactClientError(err)}`;
        return fallbackResponse(channel, this.lastError, latencyMs);
      }

      if (!isShapedDecisionResponse(parsed)) {
        this.lastStatus = "down";
        this.lastError = "malformed agent-bridge response (missing data/fallback)";
        return fallbackResponse(channel, this.lastError, latencyMs);
      }

      this.lastStatus = "up";
      this.lastError = String(parsed.error ?? "");

      // Normalise into the DecisionResponse contract; preserve any
      // backend-supplied fields and only fill in defaults for missing keys.
      return {
        data: parsed.data ?? null,
        fallback: Boolean(parsed.fallback),
        usage: parsed.usage ?? { promptTokens: 0, completionTokens: 0, cachedTokens: 0 },
        latencyMs: Number.isFinite(parsed.latencyMs) ? parsed.latencyMs : latencyMs,
        model: String(parsed.model ?? ""),
        error: String(parsed.error ?? ""),
        debug: parsed.debug ?? null,
      };
    } catch (err) {
      this.lastStatus = "down";
      this.lastError = compactClientError(err);
      return fallbackResponse(channel, this.lastError, nowMs() - started);
    } finally {
      clearTimeout(timer);
    }
  }

  /**
   * Poll the agent-bridge for adapter-side health. Returns
   * `{ healthy: boolean, lastSeen: number, error? }` — never throws.
   */
  async health() {
    if (typeof this.fetchImpl !== "function") {
      return { healthy: false, lastSeen: 0, error: "fetch unavailable" };
    }
    const url = `${this.agentBridgeUrl}/api/agent/${encodeURIComponent(this.agentId)}/health`;
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort("timeout"), Math.min(5000, this.timeoutMs));
    try {
      const resp = await this.fetchImpl(url, {
        method: "GET",
        headers: { ...this.headers },
        signal: ctrl.signal,
      });
      if (!resp.ok) {
        return { healthy: false, lastSeen: 0, error: `HTTP ${resp.status}` };
      }
      const data = await resp.json().catch(() => ({}));
      return {
        healthy: Boolean(data?.healthy ?? data?.ok ?? false),
        lastSeen: Number(data?.lastSeen ?? 0) || 0,
        ...(data?.error ? { error: String(data.error) } : {}),
      };
    } catch (err) {
      return { healthy: false, lastSeen: 0, error: compactClientError(err) };
    } finally {
      clearTimeout(timer);
    }
  }
}

export default HTTPAgentClient;

File: project_utopia/simulation/ai/llm/http_agent_client.py
"""HTTPAgentClient — pluggable AgentAdapter that talks to ``agent-bridge``.

Python port of ``HTTPAgentClient.js``. Each request is POSTed to
``/api/agent/<agentId>/decision`` with the body shape::

    { "channel", "payload", "schemaVersion", "options": { "timeoutMs" } }

and the response is normalized into a :class:`DecisionResponse`. Per the
Phase-1 brief, deep agent-bridge integration is Phase 2; we ship the class
signature + transport so the bench framework can swap it in.
"""

from __future__ import annotations

import asyncio
import json
import time
from typing import Any
from urllib.parse import quote

from .agent_adapter import (
    AgentAdapter,
    CHANNELS,
    Channel,
    DecisionResponse,
    SCHEMA_VERSION,
    UsageStats,
)

DEFAULT_BRIDGE_URL = "http://localhost:8788"
DEFAULT_AGENT_ID = "default"
DEFAULT_TIMEOUT_MS = 30_000.0


def _compact_error(err: BaseException) -> str:
    raw = str(err) or err.__class__.__name__
    msg = " ".join(raw.split())
    low = msg.lower()
    if "timeout" in low or "timed out" in low:
        return "request timeout"
    if "connection" in low and ("refused" in low or "reset" in low):
        return "agent-bridge unreachable"
    return msg[:200]


def _now_ms() -> float:
    return time.perf_counter() * 1000.0


def _trim_url(url: str) -> str:
    return (url or "").rstrip("/")


def _fallback_response(channel: Channel, error: str, latency_ms: float = 0.0,
                       debug: dict[str, Any] | None = None) -> DecisionResponse:
    return DecisionResponse(
        data=None,
        fallback=True,
        latency_ms=latency_ms,
        model="fallback",
        error=str(error or ""),
        debug=debug or {"channel": channel, "source": "http-agent-client"},
    )


class HTTPAgentClient(AgentAdapter):
    """AgentAdapter that talks to a per-agent HTTP bridge.

    Args:
        agent_bridge_url: Root URL of the bridge server.
        agent_id: Identifier appended to ``/api/agent/<agentId>/...``.
        timeout_ms: Per-request timeout in milliseconds.
        headers: Extra request headers (merged with ``Content-Type``).
        http_client: Optional injected ``httpx.AsyncClient``-style client
            for tests; falls back to lazy ``httpx.AsyncClient`` construction.
    """

    def __init__(
        self,
        *,
        agent_bridge_url: str = DEFAULT_BRIDGE_URL,
        agent_id: str = DEFAULT_AGENT_ID,
        timeout_ms: float = DEFAULT_TIMEOUT_MS,
        headers: dict[str, str] | None = None,
        http_client: Any | None = None,
    ) -> None:
        self.agent_bridge_url = _trim_url(agent_bridge_url)
        self.agent_id = str(agent_id or DEFAULT_AGENT_ID)
        self.timeout_ms = max(500.0, float(timeout_ms))
        self.headers: dict[str, str] = dict(headers or {})
        self._http_client = http_client
        self.last_error: str = ""
        self.last_latency_ms: float = 0.0
        self.last_status: str = "unknown"

    def _get_client(self) -> Any:
        if self._http_client is not None:
            return self._http_client
        try:
            import httpx  # type: ignore[import-untyped]
        except ImportError as err:
            raise RuntimeError(
                "HTTPAgentClient requires httpx; install via `pip install httpx`."
            ) from err
        self._http_client = httpx.AsyncClient(timeout=self.timeout_ms / 1000.0)
        return self._http_client

    async def aclose(self) -> None:
        client = self._http_client
        if client is not None and hasattr(client, "aclose"):
            await client.aclose()

    async def request(
        self,
        channel: Channel,
        payload: Any,
        options: dict[str, Any] | None = None,
    ) -> DecisionResponse:
        if channel not in CHANNELS:
            return _fallback_response(channel, f"unknown channel: {channel}")

        url = f"{self.agent_bridge_url}/api/agent/{quote(self.agent_id, safe='')}/decision"
        opts = options or {}
        timeout_ms = float(opts.get("timeoutMs") or self.timeout_ms)
        body = {
            "channel": channel,
            "payload": payload,
            "schemaVersion": opts.get("schemaVersion") or SCHEMA_VERSION,
            "options": {"timeoutMs": timeout_ms},
        }

        try:
            client = self._get_client()
        except RuntimeError as err:
            return _fallback_response(channel, _compact_error(err))

        started = _now_ms()
        try:
            response = await asyncio.wait_for(
                client.post(
                    url,
                    json=body,
                    headers={"Content-Type": "application/json", **self.headers},
                ),
                timeout=timeout_ms / 1000.0,
            )
        except asyncio.TimeoutError:
            latency = _now_ms() - started
            self.last_status = "down"
            self.last_error = "request timeout"
            self.last_latency_ms = latency
            return _fallback_response(channel, self.last_error, latency)
        except Exception as err:  # noqa: BLE001
            latency = _now_ms() - started
            self.last_status = "down"
            self.last_error = _compact_error(err)
            self.last_latency_ms = latency
            return _fallback_response(channel, self.last_error, latency)

        latency = _now_ms() - started
        self.last_latency_ms = latency

        if not getattr(response, "is_success", False) and getattr(response, "status_code", 200) >= 400:
            text = ""
            try:
                text = (response.text or "")[:140]
            except Exception:  # noqa: BLE001
                pass
            self.last_status = "down"
            status = getattr(response, "status_code", "?")
            self.last_error = f"HTTP {status}: {' '.join(text.split())}".strip()
            return _fallback_response(channel, self.last_error, latency)

        try:
            parsed = response.json()
        except Exception as err:  # noqa: BLE001
            self.last_status = "down"
            self.last_error = f"malformed JSON: {_compact_error(err)}"
            return _fallback_response(channel, self.last_error, latency)

        if not isinstance(parsed, dict) or (
            "data" not in parsed and "fallback" not in parsed
        ):
            self.last_status = "down"
            self.last_error = "malformed agent-bridge response (missing data/fallback)"
            return _fallback_response(channel, self.last_error, latency)

        self.last_status = "up"
        self.last_error = str(parsed.get("error") or "")

        usage_obj = parsed.get("usage")
        usage: UsageStats | None = None
        if isinstance(usage_obj, dict):
            try:
                usage = UsageStats.model_validate(usage_obj)
            except Exception:  # noqa: BLE001
                usage = None

        return DecisionResponse(
            data=parsed.get("data"),
            fallback=bool(parsed.get("fallback")),
            usage=usage,
            latency_ms=float(parsed.get("latencyMs") or latency),
            model=str(parsed.get("model") or ""),
            error=str(parsed.get("error") or ""),
            debug=parsed.get("debug") if isinstance(parsed.get("debug"), dict) else {},
        )

    async def health(self) -> dict[str, Any]:
        """Poll the agent-bridge for adapter-side health. Never raises."""
        url = f"{self.agent_bridge_url}/api/agent/{quote(self.agent_id, safe='')}/health"
        try:
            client = self._get_client()
        except RuntimeError as err:
            return {"healthy": False, "lastSeen": 0, "error": _compact_error(err)}
        try:
            response = await asyncio.wait_for(
                client.get(url, headers=dict(self.headers)),
                timeout=min(5.0, self.timeout_ms / 1000.0),
            )
        except Exception as err:  # noqa: BLE001
            return {"healthy": False, "lastSeen": 0, "error": _compact_error(err)}
        if getattr(response, "status_code", 200) >= 400:
            return {"healthy": False, "lastSeen": 0, "error": f"HTTP {response.status_code}"}
        try:
            data = response.json()
        except Exception:  # noqa: BLE001
            data = {}
        out: dict[str, Any] = {
            "healthy": bool(data.get("healthy") or data.get("ok") or False),
            "lastSeen": int(data.get("lastSeen") or 0),
        }
        if data.get("error"):
            out["error"] = str(data["error"])
        return out


__all__ = ["HTTPAgentClient", "DEFAULT_BRIDGE_URL", "DEFAULT_AGENT_ID"]
